ollama: Allow specifying a bearer token

kov opened the pull request:
Closes #19491

Release Notes:

- Added OLLAMA_API_KEY and UI to set up key for the ollama provider to allow
  for authenticated API endpoints

notpeter commented:
Hi, thanks for this PR!

The bearer token shouldn't be stored in settings as it's a secret.  We support reading API keys from the environment (e.g. ANTHROPIC_API_KEY, OPENAI_API_KEY) or keeping credentials in the platform specific keychain (search for `cx.read_credentials` and `cx.write_credentials`).

For example:
https://github.com/zed-industries/zed/blob/92c29be74cc2ac09dfe0d71d5a1048121b6ab4c6/crates/language_model/src/provider/anthropic.rs#L107-L126

Unlike the other providers, this key would be optional (e.g. local ollama requires no auth), but we'd want to support reading `OLLAMA_API_KEY` from the environment or providing a key on the Assistant configuration page like we do for OpenAI/Anthropic/Gemini.
<img width="620" alt="Screenshot 2024-10-20 at 12 52 07" src="https://github.com/user-attachments/assets/52e71e3a-77aa-4f5f-bc0d-71854b11f460">

kov committed:
ollama: allow specifying a bearer token as API key (#19491)

Release Notes:

- Added OLLAMA_API_KEY and UI to set up key for the ollama provider to allow
  for authenticated API endpoints

kov commented:
That makes sense, thank you for the quick feedback! I have started on that path. It is quite a bit more code. I got the UI to render, most of the code structure is in place, I believe, but it is still not working and the UI could use a little bit more love. I will try to get it finished during the week, may need to wait for the weekend, though.

notpeter commented:
Do you have instructions for how to configure an CloudFlare endpoint so I can test this?

kov commented:
I will take a look at clippy and tests later today. I haven't worked with CloudFlare before, unfortunately. For a quick test you could maybe use this:

https://github.com/stephan-buckmaster/ruby-bearer-auth-proxy

I set up my own nginx reverse proxy with a very simple hand rolled check, posted the config with the actual token changed below.

```
server {
    listen 0.0.0.0:443 ssl http2;
    listen [::]:443 ssl http2;
    server_name my.domain.here;
    ssl_certificate /etc/letsencrypt/live/ai.kov.dev.br/fullchain.pem; # managed by Certbot
    ssl_certificate_key /etc/letsencrypt/live/ai.kov.dev.br/privkey.pem; # managed by Certbot

    ssl_session_cache shared:SSL:1m;
    ssl_session_timeout  10m;
    ssl_protocols TLSv1 TLSv1.1 TLSv1.2;
    ssl_ciphers HIGH:SEED:!aNULL:!eNULL:!EXPORT:!DES:!RC4:!MD5:!PSK:!RSAPSK:!aDH:!aECDH:!EDH-DSS-DES-CBC3-SHA:!KRB5-DES-CBC3-SHA:!SRP;
    ssl_prefer_server_ciphers on;

    client_max_body_size 300M;

    # Check for bearer token
    set $auth_token "";
    if ($http_authorization ~* "^Bearer\s+(.*)$") {
        set $auth_token $1;
    }

    # Compare the token with the expected value
    set $valid_token 0;
    if ($auth_token = "bearer-token-goes-here") {
        set $valid_token 1;
    }

    # Return 401 Unauthorized if the token is invalid
    if ($valid_token = 0) {
        return 401;
    }

    location / {
        proxy_pass http://localhost:11434;

        proxy_headers_hash_max_size 512;
        proxy_headers_hash_bucket_size 64;
        proxy_set_header Host $host;
        proxy_set_header X-Forwarded-Proto $scheme;
        proxy_set_header X-Real-IP $remote_addr;
        proxy_set_header X-Forwarded-For $proxy_add_x_forwarded_for;

        add_header Front-End-Https on;
    }

    location = /favicon.ico {
        log_not_found off;
        access_log off;
    }
}
```

kov commented:
Hey, anything I could do to make this easier to review, or any comments you already have? Would you be comfortable letting the workflows run to see if they spot anything else?

notpeter committed:
Merge main

notpeter committed:
Merge remote-tracking branch 'origin/main' into ollama-bearer

notpeter commented:
> Hey, anything I could do to make this easier to review, or any comments you already have? Would you be comfortable letting the workflows run to see if they spot anything else?

There was a PR merged into main which result in conflicts in yours:
- https://github.com/zed-industries/zed/pull/19830

I took a stab at resolving the merge conflicts (hopefully I didn't muck it up) and my push retriggered tests.

mattgagliardi commented:
A bit of searching led me to this PR...just chucking out a quick "bump" ;)

bennetbo committed:
Merge branch 'main' into ollama-bearer

bennetbo commented:
Sorry for taking so long to take a look at this, resolved the merged conflicts for now and will follow up with a review this week.

One thing that I already noticed is that we always show the API key as "configured" when we are connected. This means that the UI shows this:
<img width="628" alt="image" src="https://github.com/user-attachments/assets/90bb6c7a-cccb-41c1-85ea-affbe4d6cef4">

even for my locally running Ollama instance, which is not using any authentication. Also, clicking on "Reset API key" does nothing in this case.

ConradIrwin commented:
Closing for now, but please do reopen when @bennetbo's comments are resolved!
